File: conformance/haskell/src/RecordedWire.hs
{-# LANGUAGE OverloadedStrings #-}
-- This is a consumer composition acceptance witness, not a record/follow API.
-- Store append and subscriber handoff share one lock. Every subscriber owns its
-- bounded writer, while the production Wire views provide selection and mounts.
module RecordedWire (recordedWireWitness) where

import Control.Concurrent (forkIO)
import Control.Concurrent.MVar
import Control.Concurrent.STM
import Control.Exception (bracket, finally, throwIO, try)
import Control.Monad (forM_, unless, when)
import Data.Aeson (Value (..), object, (.=))
import qualified Data.Aeson.KeyMap as KM
import qualified Data.Map.Strict as Map
import qualified Data.Sequence as Seq
import Data.Foldable (toList)
import Data.Text (Text)
import Bitwire hiding (errorCode, errorMessage, errorData)
import Nightseam.Runtime.WireFrames
import Nightseam.Duplex.Dispatcher
import Nightseam.Duplex.Wire
import System.Timeout (timeout)

data Entry = Entry Path Message Int

data Follower = Follower
  { followerLive :: TBQueue Entry
  , followerStop :: TVar Bool
  , followerDone :: TMVar ()
  , followerSent :: TBQueue Int
  , followerPaused :: TMVar ()
  , followerResume :: TMVar ()
  }

data StoreState = StoreState
  { storedEntries :: Seq.Seq Entry
  , storedFollowers :: Map.Map Int Follower
  , nextFollower :: Int
  }

newtype Store = Store (MVar StoreState)

newStore :: IO Store
newStore = Store <$> newMVar (StoreState Seq.empty Map.empty 0)

storeHead :: Store -> IO Int
storeHead (Store state) = withMVar state (pure . Seq.length . storedEntries)

storeWire :: Store -> Wire
storeWire (Store state) = Wire
  { send = \path message -> modifyMVar_ state $ \current -> do
      let sequenceNumber = Seq.length (storedEntries current) + 1
          entry = Entry path message sequenceNumber
      followers <- fmap Map.fromList $ fmap concat $ mapM (admit entry) (Map.toList (storedFollowers current))
      pure current {storedEntries = storedEntries current Seq.|> entry, storedFollowers = followers}
  }
  where
    admit entry (identifier, follower) = atomically $ do
      full <- isFullTBQueue (followerLive follower)
      if full then do
        -- Signal only: ending the subscriber runs in its own writer and never
        -- invokes application code while append exclusion is held.
        writeTVar (followerStop follower) True
        pure []
      else do
        writeTBQueue (followerLive follower) entry
        pure [(identifier, follower)]

closeStore :: Store -> IO ()
closeStore (Store state) = modifyMVar_ state $ \current -> do
  forM_ (Map.elems (storedFollowers current)) (\follower -> atomically (writeTVar (followerStop follower) True))
  pure current {storedFollowers = Map.empty}

attach :: Store -> Int -> Endpoint -> Int -> Bool -> IO (Int, Follower)
attach (Store state) after target bound pause = do
  follower <- Follower <$> newTBQueueIO (fromIntegral bound) <*> newTVarIO False <*> newEmptyTMVarIO
    <*> newTBQueueIO 32 <*> newEmptyTMVarIO <*> newEmptyTMVarIO
  (headValue, history) <- modifyMVar state $ \current -> do
    let headValue = Seq.length (storedEntries current)
        history = toList (Seq.drop after (storedEntries current))
        identifier = nextFollower current
    pure (current
      { storedFollowers = Map.insert identifier follower (storedFollowers current)
      , nextFollower = identifier + 1
      }, (headValue, history))
  let stopped = readTVar (followerStop follower) >>= check
      sendEntry (Entry path message sequenceNumber) = do
        stop <- readTVarIO (followerStop follower)
        if stop then pure False else do
          send (endpointWire target) path message
          atomically (writeTBQueue (followerSent follower) sequenceNumber)
          pure True
      replay _ [] = follow
      replay index (entry : rest) = do
        sent <- sendEntry entry
        when sent $ do
          continue <- if index == (0 :: Int) && pause then do
            atomically (putTMVar (followerPaused follower) ())
            atomically ((stopped >> pure False) `orElse` (readTMVar (followerResume follower) >> pure True))
          else pure True
          when continue (replay (index + 1) rest)
      follow = do
        entry <- atomically ((stopped >> pure Nothing) `orElse` (Just <$> readTBQueue (followerLive follower)))
        case entry of
          Nothing -> pure ()
          Just item -> sendEntry item >>= \sent -> when sent follow
      finish = close target (Code 1008) "recorded handoff ended" `finally` atomically (putTMVar (followerDone follower) ())
  _ <- forkIO (replay 0 history `finally` finish)
  pure (headValue, follower)

stopFollower :: Follower -> IO ()
stopFollower follower = atomically (writeTVar (followerStop follower) True) >> atomically (readTMVar (followerDone follower))

awaitSent :: Follower -> Int -> IO ()
awaitSent follower lastSequence = do
  sent <- atomically (readTBQueue (followerSent follower))
  unless (sent == lastSequence) (awaitSent follower lastSequence)

-- A fixture root supplies bounded asynchronous delivery. Its receivers may
-- reenter the store, proving callback execution is outside append exclusion.
newRoot :: IO Endpoint
newRoot = do
  receivers <- newMVar (Map.empty :: Map.Map Path Receiver)
  queue <- newTBQueueIO 16
  stopped <- newTVarIO False
  done <- newEmptyTMVarIO
  let run = do
        next <- atomically ((readTVar stopped >>= check >> pure Nothing) `orElse` (Just <$> readTBQueue queue))
        case next of
          Nothing -> pure ()
          Just (Entry path message _) -> do
            receiver <- withMVar receivers (pure . Map.lookup [])
            forM_ receiver (\held -> maybe (pure ()) (\f -> f path message) (onMessage held))
            run
  _ <- forkIO (run `finally` atomically (putTMVar done ()))
  pure Endpoint
    { endpointWire = Wire $ \path message -> atomically $ do
        closed <- readTVar stopped
        when closed (throwSTM WireClosed)
        full <- isFullTBQueue queue
        when full (throwSTM (userError "recorded witness output queue full"))
        writeTBQueue queue (Entry path message 0)
    , receive = \receiver -> modifyMVar receivers $ \current ->
        if Map.member [] current then throwIO ReceiverExists else
          pure (Map.insert [] receiver current, modifyMVar_ receivers (pure . Map.delete []))
    , close = \_ _ -> atomically (writeTVar stopped True) >> atomically (readTMVar done)
    }

data Presentation = Presentation
  { presentationWire :: Endpoint
  , presentationRoot :: Endpoint
  , presentationEnd :: Endpoint
  , presentationRoutes :: [Dispatcher]
  , presentationValues :: TQueue Int
  , presentationClosed :: TQueue Int
  , presentationErrors :: TQueue Text
  , presentationCallbacks :: TVar Int
  }

present :: Store -> IO Presentation
present store = do
  root <- newRoot
  end <- newRoot
  rootRoutes <- newDispatcher root
  endRoutes <- newDispatcher end
  endView <- selectEndpoint endRoutes ["destination"]
  destinationMount <- mount (Map.singleton "out" endView)
  destinationRoutes <- newDispatcher destinationMount
  destination <- selectEndpoint destinationRoutes ["out"]
  rootView <- selectEndpoint rootRoutes ["source"]
  inner <- mount (Map.singleton "in" rootView)
  originMount <- mount (Map.singleton "outer" inner)
  originRoutes <- newDispatcher originMount
  origin <- selectEndpoint originRoutes ["outer", "in"]
  values <- newTQueueIO
  closed <- newTQueueIO
  errors <- newTQueueIO
  callbacks <- newTVarIO 0
  _ <- receive destination (Receiver (Just (\path message -> do
      if path /= ["tick"] then atomically (writeTQueue errors "recorded destination path mismatch") else do
        -- Must acquire the same lock as append. A callback on the producer's
        -- stack under that lock deadlocks and fails the witness deadline.
        _ <- storeHead store
        case messageNumber message of
          Nothing -> atomically (writeTQueue errors "recorded destination data is not an integer")
          Just value -> atomically $ do
            modifyTVar' callbacks (+ 1)
            writeTQueue values value)) (Just (\_ _ -> pure ())))
  _ <- receive origin (Receiver (Just (send (endpointWire destination))) (Just (\code _ -> do
      _ <- storeHead store
      atomically (writeTQueue closed (unCode code)))))
  pure (Presentation origin root end [rootRoutes, endRoutes, destinationRoutes, originRoutes] values closed errors callbacks)

closePresentation :: Presentation -> IO ()
closePresentation presentation = do
  forM_ (presentationRoutes presentation) (\routes -> closeDispatcher routes (Code 1000) "done")
  close (presentationWire presentation) (Code 1000) "done"
  close (presentationRoot presentation) (Code 1000) "done"
  close (presentationEnd presentation) (Code 1000) "done"

recordedMessage :: Int -> Message
recordedMessage value = Message (frame (object ["version" .= (1 :: Int), "kind" .= ("event" :: Text), "data" .= value])) Nothing

messageNumber :: Message -> Maybe Int
messageNumber message = case frameValue (messageFrame message) of
  Object fields -> case KM.lookup "data" fields of
    Just (Number number) -> let integer = truncate number :: Int
      in if fromIntegral integer == number then Just integer else Nothing
    _ -> Nothing
  _ -> Nothing

collect :: Presentation -> IO [Int]
collect presentation = do
  send (endpointWire (presentationWire presentation)) ["tick"] (recordedMessage 0)
  let loop values = do
        next <- atomically ((Left <$> readTQueue (presentationErrors presentation)) `orElse`
          (Right <$> readTQueue (presentationValues presentation)))
        case next of
          Left problem -> fail (show problem)
          Right 0 -> pure (reverse values)
          Right value -> loop (value : values)
  loop []

headCase :: Bool -> IO Value
headCase before = bracket newStore closeStore $ \store ->
  bracket (present store) closePresentation $ \presentation -> do
    let source = at (storeWire store) []
    let appendValue value = send source ["tick"] (recordedMessage value)
    mapM_ appendValue [1 .. 3]
    when before (appendValue 4)
    bracket (attach store 0 (presentationWire presentation) 2 True) (stopFollower . snd) $ \(headValue, follower) -> do
      atomically (readTMVar (followerPaused follower))
      produced <- newEmptyTMVarIO
      _ <- forkIO (mapM_ appendValue [(if before then 5 else 4) .. 5] >> atomically (putTMVar produced ()))
      -- Producer completion is observed before releasing the paused replay.
      atomically (readTMVar produced)
      producerProgress <- atomically (isEmptyTMVar (followerResume follower))
      atomically (putTMVar (followerResume follower) ())
      awaitSent follower 5
      appendValue 6
      awaitSent follower 6
      first <- collect presentation
      firstCallbacks <- readTVarIO (presentationCallbacks presentation)
      bracket (present store) closePresentation $ \second ->
        bracket (attach store 3 (presentationWire second) 2 False) (stopFollower . snd) $ \(_, late) -> do
          awaitSent late 6
          after <- collect second
          secondCallbacks <- readTVarIO (presentationCallbacks second)
          pure (object
            [ "cut" .= (if before then "append_before_head" else "head_before_append" :: Text)
            , "head" .= headValue, "first" .= first, "after_three" .= after
            , "producer_progress" .= producerProgress
            , "callbacks_outside_append" .= (firstCallbacks == length first + 1 && secondCallbacks == length after + 1)
            ])

stallCase :: IO Value
stallCase = bracket newStore closeStore $ \store -> do
  let appendValue value = send (storeWire store) ["tick"] (recordedMessage value)
  mapM_ appendValue [1 .. 3]
  bracket (present store) closePresentation $ \stalled ->
    bracket (present store) closePresentation $ \healthy ->
      bracket (attach store 0 (presentationWire stalled) 2 True) (stopFollower . snd) $ \(_, slow) -> do
        atomically (readTMVar (followerPaused slow))
        bracket (attach store 3 (presentationWire healthy) 2 False) (stopFollower . snd) $ \(_, fast) -> do
          forM_ [4 .. 5] (\value -> appendValue value >> awaitSent fast value)
          queued <- atomically (lengthTBQueue (followerLive slow))
          appendValue 6
          awaitSent fast 6
          atomically (readTMVar (followerDone slow))
          code <- atomically (readTQueue (presentationClosed stalled))
          remainingCloses <- atomically (flushTQueue (presentationClosed stalled))
          refused <- try (send (endpointWire (presentationWire stalled)) ["tick"] (recordedMessage 99))
          unless (refused == Left WireClosed) (fail "stalled carrier accepted after close")
          underneath <- newEmptyTMVarIO
          _ <- register (head (presentationRoutes stalled)) ["probe"] (Receiver (Just (\_ message -> forM_ (messageNumber message) (atomically . putTMVar underneath))) (Just (\_ _ -> pure ())))
          send (endpointWire (presentationRoot stalled)) ["probe"] (recordedMessage 99)
          probe <- atomically (readTMVar underneath)
          appendValue 7
          awaitSent fast 7
          values <- collect healthy
          headValue <- storeHead store
          pure (object
            [ "bound" .= (2 :: Int), "queued_at_bound" .= queued
            , "closed" .= (1 + length remainingCloses), "close_code" .= code
            , "healthy" .= values, "underneath" .= [probe], "head" .= headValue
            , "producer_progress" .= (headValue == 7 && not (null values) && last values == 7)
            ])

recordedWireWitness :: IO Value
recordedWireWitness = do
  result <- timeout 5000000 $ do
    cases <- mapM headCase [False, True]
    stalled <- stallCase
    pure (object ["cases" .= cases, "stalled" .= stalled])
  maybe (fail "recorded wire witness timed out") pure result

frame :: Value -> ProfileFrame
frame = either (error . show) id . parseWireFrame

File: conformance/cpp/recorded_wire.cpp
#include "recorded_wire.hpp"
#include <nightseam/duplex/wire.hpp>
#include <atomic>
#include <condition_variable>
#include <deque>
#include <mutex>
#include <thread>

namespace nightseam::conformance {
namespace {
using namespace duplex;
using runtime::Value;

template<class T> class Channel {
    std::mutex mutex_;
    std::condition_variable_any changed_;
    std::deque<T> queue_;
    std::exception_ptr failure_;
public:
    void push(T value) {
        { std::lock_guard lock(mutex_); queue_.push_back(std::move(value)); }
        changed_.notify_all();
    }
    void fail(std::exception_ptr failure) {
        { std::lock_guard lock(mutex_); failure_ = failure; }
        changed_.notify_all();
    }
    T pop(Wait wait) {
        std::unique_lock lock(mutex_);
        if (!changed_.wait_until(lock, wait.stop, wait.deadline, [&] { return failure_ || !queue_.empty(); })) {
            wait.check(); throw DeadlineExceeded();
        }
        if (failure_) std::rethrow_exception(failure_);
        T value = std::move(queue_.front()); queue_.pop_front(); return value;
    }
    std::size_t size() { std::lock_guard lock(mutex_); return queue_.size(); }
};

struct Entry { Path path; Message message; int sequence = 0; };

// This is the witness's asynchronous fixture root, not a transport or public
// record/follow primitive. Production selection and mounting do all routing.
class Root final : public Wire, public std::enable_shared_from_this<Root> {
    std::mutex mutex_;
    std::condition_variable changed_;
    std::map<std::string, Receiver> receivers_;
    std::deque<Entry> queue_;
    bool stopped_ = false;
    std::thread worker_;
    void run() {
        for (;;) {
            Entry entry;
            Receiver receiver;
            {
                std::unique_lock lock(mutex_);
                changed_.wait(lock, [&] { return stopped_ || !queue_.empty(); });
                if (stopped_) return;
                entry = std::move(queue_.front()); queue_.pop_front();
                auto found = receivers_.find(encode_path(entry.path));
                if (found != receivers_.end()) receiver = found->second;
            }
            if (receiver.message) receiver.message(entry.path, entry.message);
        }
    }
public:
    Root() : worker_([this] { run(); }) {}
    ~Root() override { close(1000, "done"); }
    void send(const Path& path, const Message& message) override {
        encode_path(path);
        {
            std::lock_guard lock(mutex_);
            if (stopped_) throw Closed();
            if (queue_.size() >= 16) throw std::runtime_error("witness output queue full");
            queue_.push_back({path, message});
        }
        changed_.notify_one();
    }
    Detach receive(const Path& path, Receiver receiver) override {
        const auto key = encode_path(path);
        std::lock_guard lock(mutex_);
        if (stopped_) throw Closed();
        if (receivers_.contains(key)) throw ReceiverExists();
        receivers_[key] = std::move(receiver);
        return [weak = weak_from_this(), key, once = std::make_shared<std::once_flag>()] {
            std::call_once(*once, [&] {
                if (auto self = weak.lock()) { std::lock_guard lock(self->mutex_); self->receivers_.erase(key); }
            });
        };
    }
    void close(int, std::string) override {
        { std::lock_guard lock(mutex_); stopped_ = true; }
        changed_.notify_all();
        if (worker_.joinable()) worker_.join();
    }
};

class Follower {
    WirePtr target_;
    std::size_t bound_;
    std::mutex mutex_;
    std::condition_variable changed_;
    std::deque<Entry> live_;
    bool stopped_ = false;
    bool resumed_ = false;
    std::thread worker_;
    bool send(const Entry& entry) {
        { std::lock_guard lock(mutex_); if (stopped_) return false; }
        target_->send(entry.path, entry.message);
        sent.push(entry.sequence);
        return true;
    }
    void run(const std::vector<Entry>& history, bool pause) {
        try {
            bool running = true;
            for (std::size_t i = 0; i < history.size(); ++i) {
                if (!send(history[i])) { running = false; break; }
                if (i == 0 && pause) {
                    paused.push(true);
                    std::unique_lock lock(mutex_);
                    changed_.wait(lock, [&] { return stopped_ || resumed_; });
                    if (stopped_) { running = false; break; }
                }
            }
            while (running) {
                Entry entry;
                {
                    std::unique_lock lock(mutex_);
                    changed_.wait(lock, [&] { return stopped_ || !live_.empty(); });
                    if (stopped_) break;
                    entry = std::move(live_.front()); live_.pop_front();
                }
                if (!send(entry)) break;
            }
        } catch (...) { sent.fail(std::current_exception()); }
        try { target_->close(1008, "recorded handoff ended"); }
        catch (...) { done.fail(std::current_exception()); }
        done.push(true);
    }
public:
    Channel<int> sent;
    Channel<bool> paused, done;
    Follower(WirePtr target, std::size_t bound) : target_(std::move(target)), bound_(bound) {}
    ~Follower() { stop(); if (worker_.joinable()) worker_.join(); }
    void start(std::vector<Entry> history, bool pause) {
        worker_ = std::thread([this, history = std::move(history), pause] { run(history, pause); });
    }
    bool enqueue(const Entry& entry) {
        bool accepted;
        {
            std::lock_guard lock(mutex_);
            accepted = !stopped_ && live_.size() < bound_;
            if (accepted) live_.push_back(entry);
            else stopped_ = true;
        }
        // Signal only. Closing the subscriber is its own writer's job, outside
        // append exclusion and off the producer's stack.
        changed_.notify_all();
        return accepted;
    }
    void stop() { { std::lock_guard lock(mutex_); stopped_ = true; } changed_.notify_all(); }
    void resume() { { std::lock_guard lock(mutex_); resumed_ = true; } changed_.notify_all(); }
    std::size_t queued() { std::lock_guard lock(mutex_); return live_.size(); }
    void await_sent(int last, Wait wait) { while (sent.pop(wait) != last) {} }
};

class Store final : public Wire {
    std::mutex mutex_;
    std::vector<Entry> entries_;
    std::vector<std::weak_ptr<Follower>> followers_;
public:
    int head() { std::lock_guard lock(mutex_); return static_cast<int>(entries_.size()); }
    void send(const Path& path, const Message& message) override {
        std::lock_guard lock(mutex_);
        Entry entry{path, message, static_cast<int>(entries_.size()) + 1};
        entries_.push_back(entry);
        for (auto i = followers_.begin(); i != followers_.end();) {
            auto follower = i->lock();
            if (!follower || !follower->enqueue(entry)) i = followers_.erase(i);
            else ++i;
        }
    }
    Detach receive(const Path&, Receiver) override { throw NoRoute(); }
    void close(int, std::string) override {
        std::vector<std::weak_ptr<Follower>> followers;
        { std::lock_guard lock(mutex_); followers.swap(followers_); }
        for (const auto& weak : followers) if (auto follower = weak.lock()) follower->stop();
    }
    std::pair<int, std::shared_ptr<Follower>> attach(int after, WirePtr target, std::size_t bound, bool pause) {
        auto follower = std::make_shared<Follower>(std::move(target), bound);
        std::vector<Entry> history;
        int head;
        {
            std::lock_guard lock(mutex_);
            head = static_cast<int>(entries_.size());
            history.assign(entries_.begin() + after, entries_.end());
            // The head and registration share append's lock, preventing a gap
            // or duplicate at the replay-to-follow cut.
            followers_.push_back(follower);
        }
        follower->start(std::move(history), pause);
        return {head, follower};
    }
};

Message tick(int value) {
    Message message;
    message.frame.kind = ProfileKind::event;
    message.frame.data = std::to_string(value);
    return message;
}

struct Presentation {
    struct Observed {
        Channel<int> values, closed;
        std::atomic<int> callbacks{0};
    };
    std::shared_ptr<Root> root = std::make_shared<Root>();
    std::shared_ptr<Root> end = std::make_shared<Root>();
    std::shared_ptr<Observed> observed = std::make_shared<Observed>();
    WirePtr wire, destination;
    explicit Presentation(const std::shared_ptr<Store>& store) {
        destination = at(mount({{"out", at(end, {"destination"})}}), {"out"});
        wire = at(mount({{"outer", mount({{"in", at(root, {"source"})}})}}), {"outer", "in"});
        destination->receive({"tick"}, {false, [store, state = observed](const Path& path, const Message& message) {
            try {
                if (path != Path{"tick"}) throw std::runtime_error("recorded destination received wrong relative path");
                // This real callback reenters append. Calling it under append
                // exclusion deadlocks and fails the witness's deadline.
                (void)store->head();
                ++state->callbacks;
                state->values.push(runtime::parse_value(message.frame.data.value()).as<int>());
            } catch (...) { state->values.fail(std::current_exception()); }
        }, {}});
        wire->receive({"tick"}, {false, [target = destination, state = observed](const Path& path, const Message& message) {
            // An opaque forwarding hop, not payload interpretation.
            try { target->send(path, message); }
            catch (...) { state->values.fail(std::current_exception()); }
        }, [store, state = observed](int code, const std::string&) {
            (void)store->head(); state->closed.push(code);
        }});
    }
    ~Presentation() {
        wire->close(1000, "done");
        root->close(1000, "done");
        destination->close(1000, "done");
        end->close(1000, "done");
    }
    Value collect(Wait wait) {
        // A fence follows the same nested mount and forward route, so delayed
        // duplicates cannot hide behind a fixed expected delivery count.
        wire->send({"tick"}, tick(0));
        Value values = Value::array();
        for (;;) {
            const auto value = observed->values.pop(wait);
            if (value == 0) return values;
            values.push_back(value);
        }
    }
};

Value head_case(Wait wait, bool before) {
    auto store = std::make_shared<Store>();
    Presentation first(store);
    auto source = at(mount({{"record", store}}), {"record"});
    for (int value = 1; value <= 3; ++value) source->send({"tick"}, tick(value));
    const auto cut = before ? "append_before_head" : "head_before_append";
    int next = 4;
    if (before) { source->send({"tick"}, tick(4)); next = 5; }
    auto [head, follower] = store->attach(0, first.wire, 2, true);
    follower->paused.pop(wait);
    Channel<bool> produced;
    std::jthread producer([&] {
        try {
            for (int value = next; value <= 5; ++value) source->send({"tick"}, tick(value));
            produced.push(true);
        } catch (...) { produced.fail(std::current_exception()); }
    });
    const bool progress = produced.pop(wait); // Writer is still held here.
    follower->resume(); follower->await_sent(5, wait);
    source->send({"tick"}, tick(6)); follower->await_sent(6, wait);
    auto values = first.collect(wait);
    Presentation second(store);
    auto [late_head, late] = store->attach(3, second.wire, 2, false);
    (void)late_head;
    late->await_sent(6, wait);
    auto after = second.collect(wait);
    Value result;
    result["cut"] = cut; result["head"] = head;
    result["first"] = std::move(values); result["after_three"] = std::move(after);
    result["producer_progress"] = progress;
    result["callbacks_outside_append"] = first.observed->callbacks.load() > 0 && second.observed->callbacks.load() > 0;
    return result;
}

Value stall_case(Wait wait) {
    auto store = std::make_shared<Store>();
    for (int value = 1; value <= 3; ++value) store->send({"tick"}, tick(value));
    Presentation stalled(store), healthy(store);
    auto [head, slow] = store->attach(0, stalled.wire, 2, true);
    (void)head;
    slow->paused.pop(wait);
    auto [healthy_head, fast] = store->attach(3, healthy.wire, 2, false);
    (void)healthy_head;
    for (int value = 4; value <= 5; ++value) {
        store->send({"tick"}, tick(value)); fast->await_sent(value, wait);
    }
    const auto queued = slow->queued();
    store->send({"tick"}, tick(6)); fast->await_sent(6, wait);
    slow->done.pop(wait);
    const auto code = stalled.observed->closed.pop(wait);
    bool refused = false;
    try { stalled.wire->send({"tick"}, tick(99)); } catch (const Closed&) { refused = true; }
    if (!refused) throw std::runtime_error("stalled carrier accepted after close");
    Channel<int> underneath;
    stalled.root->receive({"probe"}, {false, [&](const Path&, const Message& message) {
        try { underneath.push(runtime::parse_value(message.frame.data.value()).as<int>()); }
        catch (...) { underneath.fail(std::current_exception()); }
    }, {}});
    stalled.root->send({"probe"}, tick(99));
    const auto probe = underneath.pop(wait);
    store->send({"tick"}, tick(7)); fast->await_sent(7, wait);
    auto values = healthy.collect(wait);
    Value result;
    result["bound"] = 2; result["queued_at_bound"] = queued;
    result["closed"] = 1 + stalled.observed->closed.size(); result["close_code"] = code;
    result["healthy"] = std::move(values);
    Value probes = Value::array(); probes.push_back(probe); result["underneath"] = std::move(probes);
    result["head"] = store->head(); result["producer_progress"] = true;
    return result;
}
} // namespace

Value recorded_wire_witness(duplex::Wait wait) {
    wait.check();
    Value cases = Value::array();
    cases.push_back(head_case(wait, false)); cases.push_back(head_case(wait, true));
    Value result;
    result["cases"] = std::move(cases); result["stalled"] = stall_case(wait);
    return result;
}

} // namespace nightseam::conformance
